Avoid changing the ordering of attributes when markup is not created for saving page

szewai opened the pull request:
#### d10b5ad2c28f6657737d54517e5d8486568c35c9
<pre>
Avoid changing the ordering of attributes when markup is not created for saving page
<a href="https://bugs.webkit.org/show_bug.cgi?id=273130">https://bugs.webkit.org/show_bug.cgi?id=273130</a>
<a href="https://rdar.apple.com/126120490">rdar://126120490</a>

Reviewed by Ryosuke Niwa.

URL replacement mode in MarkupAccumulator is introduced specifically for saving web page, and it should not affect
result of other cases; otherwise it might break existing usecases. 276043@main accidentally changes the ordering of
attributes by appending crossorigin attribute to the end -- even when MarkupAccumulator is not created with URL
replacement mode. This patch fixes that by appending attributes in original order when URL replacement is not enabled.
(see MarkupAccumulator::appendStartTag).

Also, to make it easier to distinguish between URL replacement mode and others, this patch adds two functions:
serializeFragmentWithURLReplacement and MarkupAccumulator::enableURLReplacement.

* LayoutTests/http/tests/dom/resources/simple-iframe.html: Added.
* LayoutTests/http/tests/dom/serialize-fragment-attributes-order-expected.txt: Added.
* LayoutTests/http/tests/dom/serialize-fragment-attributes-order.html: Added.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::getHTML const):
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::getHTML const):
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::MarkupAccumulator):
(WebCore::MarkupAccumulator::enableURLReplacement):
(WebCore::MarkupAccumulator::appendContentsForNode):
(WebCore::MarkupAccumulator::resolveURLIfNeeded const):
(WebCore::MarkupAccumulator::appendStartTagWithURLReplacement):
(WebCore::MarkupAccumulator::appendStartTag):
(WebCore::MarkupAccumulator::frameForAttributeReplacement const):
(WebCore::MarkupAccumulator::replaceAttributeIfNecessary):
(WebCore::MarkupAccumulator::appendURLAttributeForReplacementIfNecessary):
* Source/WebCore/editing/MarkupAccumulator.h:
(WebCore::MarkupAccumulator::MarkupAccumulator):
* Source/WebCore/editing/markup.cpp:
(WebCore::serializeFragment):
(WebCore::serializeFragmentWithURLReplacement):
* Source/WebCore/editing/markup.h:
(WebCore::serializeFragment):
(WebCore::serializeFragmentWithURLReplacement):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):

Canonical link: <a href="https://commits.webkit.org/277944@main">https://commits.webkit.org/277944@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e6e4348fc8d6ed70a329cb50c8c9369db5a5f40e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28199 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51953 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51674 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/45056 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51291 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34166 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25727 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40043 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49945 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25846 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42236 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21151 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23304 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43405 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7041 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45230 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43910 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53584 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24037 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20291 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47354 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25300 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42444 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46320 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10787 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26109 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25020 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/http/tests/dom/serialize-fragment-attributes-order.html:
```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE html>
+<html>
+<body>
+<script src="/js-test-resources/js-test-pre.js"></script>
```
Should use `js-test.js`

cdumez commented on LayoutTests/http/tests/dom/serialize-fragment-attributes-order.html:
```diff
@@ -0,0 +1,28 @@
+document.body.appendChild(frame);
+
+</script>
+<script src="/js-test-resources/js-test-post.js"></script>
```
Not needed

cdumez commented on LayoutTests/http/tests/dom/serialize-fragment-attributes-order.html:
```diff
@@ -0,0 +1,28 @@
+<script>
+description("Verifies the order of attributes in markup is the same as the order they are added");
+
+if (window.testRunner) {
```
All of this `if` scope should be replaced with:
```
jsTestIsAsync = true;
```

We should never use waitUntilDone() / notifyDone() with js-test.js.

cdumez commented on LayoutTests/http/tests/dom/serialize-fragment-attributes-order.html:
```diff
@@ -0,0 +1,28 @@
+frame.onload = () => {
+    shouldBeEqualToString("frame.outerHTML", '<iframe id="frameID" crossorigin="anonymous" src="http://localhost:8000/dom/resources/simple-iframe.html"></iframe>');
+
+    if (window.testRunner)
```
Should call `finishJSTest()` instead since this is a js-test.js test.

szewai committed:
Avoid changing the ordering of attributes when markup is not created for saving page
https://bugs.webkit.org/show_bug.cgi?id=273130
rdar://126120490

Reviewed by Ryosuke Niwa.

URL replacement mode in MarkupAccumulator is introduced specifically for saving web page, and it should not affect
result of other cases; otherwise it might break existing usecases. 276043@main accidentally changes the ordering of
attributes by appending crossorigin attribute to the end -- even when MarkupAccumulator is not created with URL
replacement mode. This patch fixes that by appending attributes in original order when URL replacement is not enabled.
(see MarkupAccumulator::appendStartTag).

Also, to make it easier to distinguish between URL replacement mode and others, this patch adds two functions:
serializeFragmentWithURLReplacement and MarkupAccumulator::enableURLReplacement.

* LayoutTests/http/tests/dom/resources/simple-iframe.html: Added.
* LayoutTests/http/tests/dom/serialize-fragment-attributes-order-expected.txt: Added.
* LayoutTests/http/tests/dom/serialize-fragment-attributes-order.html: Added.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::getHTML const):
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::getHTML const):
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::MarkupAccumulator):
(WebCore::MarkupAccumulator::enableURLReplacement):
(WebCore::MarkupAccumulator::appendContentsForNode):
(WebCore::MarkupAccumulator::resolveURLIfNeeded const):
(WebCore::MarkupAccumulator::appendStartTagWithURLReplacement):
(WebCore::MarkupAccumulator::appendStartTag):
(WebCore::MarkupAccumulator::frameForAttributeReplacement const):
(WebCore::MarkupAccumulator::replaceAttributeIfNecessary):
(WebCore::MarkupAccumulator::appendURLAttributeForReplacementIfNecessary):
* Source/WebCore/editing/MarkupAccumulator.h:
(WebCore::MarkupAccumulator::MarkupAccumulator):
* Source/WebCore/editing/markup.cpp:
(WebCore::serializeFragment):
(WebCore::serializeFragmentWithURLReplacement):
* Source/WebCore/editing/markup.h:
(WebCore::serializeFragment):
(WebCore::serializeFragmentWithURLReplacement):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):

Canonical link: https://commits.webkit.org/277944@main
